File: src/modules/profile/dto/company-profile.dto.ts
import {
  IsNotEmpty,
  IsString,
  IsEnum,
  IsOptional,
  IsArray,
  IsNumber,
  IsDate,
  Min,
} from 'class-validator';
import { Type } from 'class-transformer';
import { ApiProperty, ApiPropertyOptional, PartialType } from '@nestjs/swagger';

// Enums
export enum ContactType {
  MOBILE = 'mobile',
  EMAIL = 'email',
  WEBSITE = 'website',
  LINKEDIN = 'linkedin',
  INSTAGRAM = 'instagram',
  FACEBOOK = 'facebook',
}

export enum AddressType {
  RESIDENTIAL = 'residential',
  COMMERCIAL = 'commercial',
}

export enum AccountType {
  CURRENT = 'currentAccount',
  SAVINGS = 'savingsAccount',
}

// Nested DTOs
export class PartnerDto {
  @ApiProperty({
    example: 'personId',
    description: 'Reference to person',
  })
  @IsNotEmpty()
  @IsString()
  personId: string;
}

export class ContactDto {
  @ApiProperty({ enum: ContactType, example: ContactType.MOBILE })
  @IsNotEmpty()
  @IsEnum(ContactType)
  contactType: ContactType;

  @ApiProperty({ example: '+55 11 91234-5678' })
  @IsNotEmpty()
  @IsString()
  contactValue: string;

  @ApiProperty({ example: 'WhatsApp' })
  @IsNotEmpty()
  @IsString()
  contactComplement: string;
}

export class RelatedFileDto {
  @ApiProperty({ example: 'Descrição do arquivo' })
  @IsNotEmpty()
  @IsString()
  filesDescription: string;

  @ApiProperty({ example: 'arquivo.pdf' })
  @IsNotEmpty()
  @IsString()
  relatedFilesFiles: string;

  @ApiProperty({ example: 20 })
  @IsNotEmpty()
  @IsNumber()
  relatedFilesDateDay: number;

  @ApiProperty({ example: 4 })
  @IsNotEmpty()
  @IsNumber()
  relatedFilesDateMonth: number;

  @ApiProperty({ example: 2023 })
  @IsNotEmpty()
  @IsNumber()
  relatedFilesDateYear: number;
}

export class BankDataDto {
  @ApiPropertyOptional({ example: '341' })
  @IsOptional()
  @IsString()
  bankName?: string;

  @ApiPropertyOptional({ example: '1234' })
  @IsOptional()
  @IsString()
  bankBranch?: string;

  @ApiPropertyOptional({ example: '56789-0' })
  @IsOptional()
  @IsString()
  bankAccount?: string;

  @ApiPropertyOptional({ enum: AccountType, example: AccountType.CURRENT })
  @IsOptional()
  @IsEnum(AccountType)
  bankAccountType?: AccountType;
}

export class CreateCompanyProfileDto {
  @ApiProperty({ example: 'userId', readOnly: true })
  @IsNotEmpty()
  @IsString()
  userId: string;

  @ApiProperty({ example: '12.345.678/0001-90' })
  @IsOptional()
  @IsString()
  cnpj?: string;

  @ApiProperty({ example: 'Empresa ABC LTDA' })
  @IsNotEmpty()
  @IsString()
  companyName: string;

  @ApiProperty({ example: 'ABC Negócios' })
  @IsNotEmpty()
  @IsString()
  businessName: string;

  @ApiPropertyOptional({
    example: '2001-01-01',
    type: 'string',
    format: 'date',
  })
  @IsOptional()
  @IsDate()
  @Type(() => Date)
  birthday?: Date;

  @ApiPropertyOptional({ example: 'Sociedade Limitada' })
  @IsOptional()
  @IsString()
  legalNature?: string;

  @ApiPropertyOptional({ example: 'Empresa especializada em...' })
  @IsOptional()
  @IsString()
  companyDescription?: string;

  @ApiPropertyOptional({ example: 'logo.png' })
  @IsOptional()
  @IsString()
  logoImage?: string;

  @ApiPropertyOptional({ example: ['foto1.png', 'foto2.png'] })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  companyImages?: string[];

  @ApiProperty({
    description: 'Related tag IDs',
    example: ['tagId1', 'tagId2'],
    isArray: true,
  })
  @IsNotEmpty()
  @IsArray()
  @IsString({ each: true })
  tagId: string[];

  // Sócios
  @ApiPropertyOptional({ type: [PartnerDto] })
  @IsOptional()
  @IsArray()
  @Type(() => PartnerDto)
  partners?: PartnerDto[];

  // Contatos
  @ApiPropertyOptional({ type: [ContactDto] })
  @IsOptional()
  @IsArray()
  @Type(() => ContactDto)
  contacts?: ContactDto[];

  // Endereço 1
  @ApiPropertyOptional({ example: '01001-000' })
  @IsOptional()
  @IsString()
  addressOneCepBrasilApi?: string;

  @ApiPropertyOptional({ enum: AddressType, example: AddressType.COMMERCIAL })
  @IsOptional()
  @IsString()
  addressOneType?: string;

  @ApiPropertyOptional({ example: 'Av. Paulista' })
  @IsOptional()
  @IsString()
  addressOneStreet?: string;

  @ApiPropertyOptional({ example: '1000' })
  @IsOptional()
  @IsString()
  addressOneNumber?: string;

  @ApiPropertyOptional({ example: 'Fundos' })
  @IsOptional()
  @IsString()
  addressOneComplement?: string;

  @ApiPropertyOptional({ example: 'São Paulo' })
  @IsOptional()
  @IsString()
  addressOneCity?: string;

  @ApiPropertyOptional({ example: 'SP' })
  @IsOptional()
  @IsString()
  addressOneState?: string;

  // Endereço 2
  @ApiPropertyOptional({ example: '02002-000' })
  @IsOptional()
  @IsString()
  addressTwoCepBrasilApi?: string;

  @ApiPropertyOptional({ enum: AddressType, example: AddressType.RESIDENTIAL })
  @IsOptional()
  @IsString()
  addressTwoType?: string;

  @ApiPropertyOptional({ example: 'Rua das Flores' })
  @IsOptional()
  @IsString()
  addressTwoStreet?: string;

  @ApiPropertyOptional({ example: '222' })
  @IsOptional()
  @IsString()
  addressTwoNumber?: string;

  @ApiPropertyOptional({ example: 'Casa' })
  @IsOptional()
  @IsString()
  addressTwoComplement?: string;

  @ApiPropertyOptional({ example: 'Campinas' })
  @IsOptional()
  @IsString()
  addressTwoCity?: string;

  @ApiPropertyOptional({ example: 'SP' })
  @IsOptional()
  @IsString()
  addressTwoState?: string;

  // Bancos
  @ApiPropertyOptional({ type: BankDataDto })
  @IsOptional()
  @Type(() => BankDataDto)
  bankDataOne?: BankDataDto;

  @ApiPropertyOptional({ type: BankDataDto })
  @IsOptional()
  @Type(() => BankDataDto)
  bankDataTwo?: BankDataDto;

  // Arquivos
  @ApiPropertyOptional({ type: [RelatedFileDto] })
  @IsOptional()
  @IsArray()
  @Type(() => RelatedFileDto)
  relatedFiles?: RelatedFileDto[];

  @ApiProperty({ example: 'userId', readOnly: true })
  @IsNotEmpty()
  @IsString()
  createdBy: string;

  @ApiProperty({ example: 'userId', readOnly: true })
  @IsNotEmpty()
  @IsString()
  ownerId: string;
}

export class UpdateCompanyProfileDto extends PartialType(
  CreateCompanyProfileDto,
) {}

export class CompanyProfileFilterDto {
  @ApiPropertyOptional({ example: 1 })
  @IsNumber()
  @Min(1)
  @IsOptional()
  page?: number;

  @ApiPropertyOptional({ example: 10 })
  @IsNumber()
  @Min(1)
  @IsOptional()
  limit?: number;
}

File: RapidaQuickstart.DotNet/DTOs/CompanyProfileDto.cs
using System.ComponentModel.DataAnnotations;

namespace RapidaQuickstart.DotNet.DTOs
{
    public class CreateCompanyProfileDto
    {
        [Required]
        public string UserId { get; set; } = string.Empty;

        [StringLength(18)]
        public string? Cnpj { get; set; }

        [Required]
        [StringLength(200)]
        public string CompanyName { get; set; } = string.Empty;

        [Required]
        [StringLength(200)]
        public string BusinessName { get; set; } = string.Empty;

        public DateTime? Birthday { get; set; }

        [StringLength(100)]
        public string? LegalNature { get; set; }

        [StringLength(1000)]
        public string? CompanyDescription { get; set; }

        public string? LogoImage { get; set; }

        public List<string>? CompanyImages { get; set; }

        [Required]
        public List<string> TagId { get; set; } = new();

        // Sócios
        public List<PartnerDto>? Partners { get; set; }

        // Contatos
        public List<ContactDto>? Contacts { get; set; }

        // Endereço 1
        [StringLength(9)]
        public string? AddressOneCepBrasilApi { get; set; }

        public string? AddressOneType { get; set; }

        [StringLength(200)]
        public string? AddressOneStreet { get; set; }

        [StringLength(20)]
        public string? AddressOneNumber { get; set; }

        [StringLength(100)]
        public string? AddressOneComplement { get; set; }

        [StringLength(100)]
        public string? AddressOneCity { get; set; }

        [StringLength(2)]
        public string? AddressOneState { get; set; }

        // Endereço 2
        [StringLength(9)]
        public string? AddressTwoCepBrasilApi { get; set; }

        public string? AddressTwoType { get; set; }

        [StringLength(200)]
        public string? AddressTwoStreet { get; set; }

        [StringLength(20)]
        public string? AddressTwoNumber { get; set; }

        [StringLength(100)]
        public string? AddressTwoComplement { get; set; }

        [StringLength(100)]
        public string? AddressTwoCity { get; set; }

        [StringLength(2)]
        public string? AddressTwoState { get; set; }

        // Dados Bancários
        public BankDataDto? BankDataOne { get; set; }

        public BankDataDto? BankDataTwo { get; set; }
    }

    public class UpdateCompanyProfileDto : CreateCompanyProfileDto
    {
        // Herda todos os campos do CreateCompanyProfileDto
    }

    public class CompanyProfileFilterDto
    {
        public string? Search { get; set; }
        public string? LegalNature { get; set; }
        public List<string>? TagIds { get; set; }
        public int Page { get; set; } = 1;
        public int Limit { get; set; } = 10;
        public string? SortBy { get; set; } = "createdAt";
        public string? SortDirection { get; set; } = "desc";
    }

    public class PartnerDto
    {
        [Required]
        [StringLength(100)]
        public string PartnerName { get; set; } = string.Empty;

        [Required]
        [StringLength(20)]
        public string PartnerDocument { get; set; } = string.Empty;

        [EmailAddress]
        [StringLength(100)]
        public string? PartnerEmail { get; set; }

        [StringLength(20)]
        public string? PartnerPhone { get; set; }

        [Range(0, 100)]
        public decimal ParticipationPercentage { get; set; }

        public bool IsMainPartner { get; set; } = false;
    }

    public class ContactDto
    {
        [Required]
        [StringLength(50)]
        public string ContactType { get; set; } = string.Empty; // email, phone, whatsapp, etc.

        [Required]
        [StringLength(100)]
        public string ContactValue { get; set; } = string.Empty;

        public bool IsPrimary { get; set; } = false;

        [StringLength(200)]
        public string? Description { get; set; }
    }
}
